File: jscomp/test/demo_page.ml
let rec fib = function
  | 1 | 2 -> 1
  | n -> fib (n - 1 )  + fib (n - 2)
(** Imperative style *)
let sum n =
    let v  = ref 0 in
    for i = 0 to n do
       v := !v + i
    done;
    !v
(** List map *)
type 'a list =
  | Nil
  | Cons of 'a * 'a list

let rec map f = function
  | Nil -> Nil
  | Cons (x,xs) ->  Cons (f x, map f xs)

(** Test curry and uncurry calling convention *)
let test_curry x  y =  x + y
let f = test_curry 32

(** Create a typed binding for react *)
type t
type element
external document :  t = "document" [@@bs.val ] 
external getElementById : t -> string -> element = "getElementById" [@@bs.send ]

(** Phantom types *)
type config
type component
type attrs
type component_class

external config :
      ?display_name:string ->
        render:(unit -> component) -> unit ->
          config = "" [@@bs.obj ] (** make a json object *)
external attrs:
        ?alt: string ->
        ?autoPlay: bool ->
          unit -> attrs = "" [@@bs.obj]
external str : string -> component = "%identity"

type vdom 
external vdom : vdom = "DOM" [@@bs.module "react"] [@@bs.val]


(* FIXME: investigate 
   cases:
   {[
     [@@bs.module "package1" "same_name"]
     [@@bs.module "package2" "same_name"]
   ]}
   {[
     [@@bs.module "package" "name1"]
     [@@bs.module "package" "name2"]
   ]}
*)
external h1 : vdom -> ?attrs:attrs -> component array  -> component = "h1" 
    [@@bs.send]  [@@bs.splice]
external h2 : vdom -> ?attrs:attrs -> component array  -> component = "h2" 
    [@@bs.send]  [@@bs.splice]

external h3 : vdom ->  ?attrs:attrs -> component array  -> component = "h3"
    [@@bs.send]  [@@bs.splice]

external h4 : vdom ->  ?attrs:attrs -> component array  -> component = "h4"
    [@@bs.send]  [@@bs.splice]

external div : vdom -> ?attrs:attrs -> component array ->  component = "div"
    [@@bs.send]  [@@bs.splice]

external createClass :
      config -> component_class = "createClass"
        [@@bs.val "createClass"]
        [@@bs.module "react"]
external render : component_class -> element -> unit = ""
    [@@bs.val "render"]
    [@@bs.module "react-dom"]
;;
(** Do the rendering *)
render (
     createClass (
     (config
       ~render:(fun _ ->
         div vdom
              ~attrs:(attrs ~alt:"pic" ())
              [|
                h1 vdom [| str "hello react"|];
                h2 vdom [| str "type safe!" |];
  |]
               )
        ()))) (getElementById document "hi")

File: jscomp/test/demo_page.js
'use strict';

var Curry = require("../../lib/js/curry.js");
var React = require("react");
var ReactDom = require("react-dom");

function fib(n) {
  if (n === 2 || n === 1) {
    return 1;
  } else {
    return fib(n - 1 | 0) + fib(n - 2 | 0) | 0;
  }
}

function sum(n) {
  var v = 0;
  for(var i = 0; i <= n; ++i){
    v = v + i | 0;
  }
  return v;
}

function map(f, param) {
  if (param) {
    return /* Cons */[
            Curry._1(f, param[0]),
            map(f, param[1])
          ];
  } else {
    return /* Nil */0;
  }
}

function test_curry(x, y) {
  return x + y | 0;
}

function f(param) {
  return 32 + param | 0;
}

ReactDom.render(React.createClass({
          render: (function () {
              return React.DOM.div({
                          alt: "pic"
                        }, React.DOM.h1(undefined, "hello react"), React.DOM.h2(undefined, "type safe!"));
            })
        }), document.getElementById("hi"));

exports.fib = fib;
exports.sum = sum;
exports.map = map;
exports.test_curry = test_curry;
exports.f = f;
/*  Not a pure module */
